### PR TITLE
Create submission job

### PR DESCRIPTION
**Pull request by matthewdu**
We have a job queue

### CODE REVIEW

File: config/application.rb [hunk 1]
```diff
@@ -26,6 +26,8 @@ class Application < Rails::Application
     # Skip views, helpers and assets when generating a new resource.
     config.api_only = true
 
+    # config.active_job.queue_adapter = :sidekiq
```
Comment (jchorl) [inline config/application.rb]: What is this? Why is it commented?

Comment (matthewdu) [inline config/application.rb]: That's to determine which backend to use for the queue implementation. sidekiq a worker/queue that backs onto redis. There are also synchronous queues, db backed queues, etc. I'm still thinking about which queue implementation to use.

File: app/jobs/run_submissions_job.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+class RunSubmissionsJob < ApplicationJob
+  queue_as :submission
```
Comment (rwongone) [inline app/jobs/run_submissions_job.rb]: To differentiate this from the Submission model, could we call the queue something like `:pending_submission`?

File: app/jobs/run_submissions_job.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    submissions.each do |submission|
+      run_submission submission
+    end
+  end
+
+  def run_submission(submission)
+    tests = Test.where(problem_id: submission.problem_id)
+
+    tests.each do |test|
+      execution = Execution.find_or_initialize_by(submission_id: submission.id, test_id: test.id)
```
Comment (rwongone) [inline app/jobs/run_submissions_job.rb]: If we're going to make Execution a relationship between Submission and Test, we need to remember not to allow updates of Tests, or Executions could be based on stale test data.

File: app/jobs/run_submissions_job.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      
+      if test.expected_output.nil?
```
Comment (rwongone) [inline app/jobs/run_submissions_job.rb]: We might want to use this PR to decide what responsibilities the job has versus the library. The library has a function that also performs this check, IIRC.

File: app/models/submission.rb [hunk 5]
```diff
@@ -1,5 +1,7 @@
 class Submission < ApplicationRecord
   has_one :db_file, foreign_key: "id", primary_key: "db_file_id"
+  has_one :execution
```
Comment (rwongone) [inline app/models/submission.rb]: Can't a Submission be executed on many tests?

File: app/jobs/run_submissions_job.rb (hunk 3 above)
Comment (matthewdu) [inline app/jobs/run_submissions_job.rb]: I was thinking that if a Test gets updated, we rerun it on all it's submissions

Comment (rwongone) [inline app/jobs/run_submissions_job.rb]: That could work too.

File: app/jobs/run_submissions_job.rb (hunk 4 above)
Comment (matthewdu) [inline app/jobs/run_submissions_job.rb]: Yeah, I was also thinking that the library should be purely functional(not modify state). Didn't want to make too many changes in this PR. I'll make another PR and come back to this once it's fixed.

**Comment by rwongone** [comment]
Maybe we should include a spec that creates Executions for 2 submissions on 2 tests.

**Comment by rwongone** [comment]
LGTM after we have a spec for Tests and Submissions with multiple Executions.

**Pull request merged**